core/templates/hash_map2.h

```diff
@@ -0,0 +1,543 @@
+		uint32_t distance = 0;
+		uint32_t pos = hash % capacity;
+
+		while (true) {
+			if (hashes[pos] == EMPTY_HASH) {
+				elements[pos] = value;
+				hashes[pos] = hash;
+
+				num_elements++;
+
+				return;
+			}
+
+			// not an empty slot, let's check the probing length of the existing one
+			uint32_t existing_probe_len = _get_probe_length(pos, hashes[pos]);
+			if (existing_probe_len < distance) {
+				SWAP(hash, hashes[pos]);
+				SWAP(value, elements[pos]);
+				distance = existing_probe_len;
+			}
+
+			pos = (pos + 1) % capacity;
+			distance++;
+		}
+	}
+
+	void _resize_and_rehash(uint32_t p_new_capacity) {
+		uint32_t old_capacity = capacity;
+
+		// Capacity can't be 0.
+		capacity = MAX(MIN_CAPACITY, p_new_capacity);
+
+		HashMap2Element<TKey, TValue> **old_elements = elements;
+		uint32_t *old_hashes = hashes;
+
+		num_elements = 0;
+		hashes = reinterpret_cast<uint32_t *>(Memory::alloc_static(sizeof(uint32_t) * capacity));
+		elements = reinterpret_cast<HashMap2Element<TKey, TValue> **>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue> *) * capacity));
+
+		for (uint32_t i = 0; i < capacity; i++) {
+			hashes[i] = 0;
+			elements[i] = nullptr;
+		}
+
+		if (old_capacity == 0) {
+			// Nothing to do.
+			return;
+		}
+
+		for (uint32_t i = 0; i < old_capacity; i++) {
+			if (old_hashes[i] == EMPTY_HASH) {
+				continue;
+			}
+
+			_insert_with_hash(old_hashes[i], old_elements[i]);
+		}
+
+		Memory::free_static(old_elements);
+		Memory::free_static(old_hashes);
+	}
+
+	void _resize_and_rehash() {
+		_resize_and_rehash(capacity * 2);
+	}
+
+	_FORCE_INLINE_ HashMap2Element<TKey, TValue> *_insert(const TKey &p_key, const TValue &p_value, bool p_front_insert = false) {
+		if (unlikely(elements == nullptr)) {
+			// Allocate on demand to save memory
+			hashes = reinterpret_cast<uint32_t *>(Memory::alloc_static(sizeof(uint32_t) * capacity));
+			elements = reinterpret_cast<HashMap2Element<TKey, TValue> **>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue> *) * capacity));
+
+			for (uint32_t i = 0; i < capacity; i++) {
+				hashes[i] = EMPTY_HASH;
+				elements[i] = nullptr;
+			}
+		}
+
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+
+		if (exists) {
+			elements[pos]->data.value = p_value;
+			return elements[pos];
+		} else {
+			if (num_elements + 1 > 0.9 * capacity) {
+				_resize_and_rehash();
+			}
+
+			HashMap2Element<TKey, TValue> *elem = element_alloc.new_allocation(HashMap2Element<TKey, TValue>(p_key, p_value));
+
+			if (tail_element == nullptr) {
+				head_element = elem;
+				tail_element = elem;
+			} else if (p_front_insert) {
+				head_element->prev = elem;
+				elem->next = head_element;
+				head_element = elem;
+			} else {
+				tail_element->next = elem;
+				elem->prev = tail_element;
+				tail_element = elem;
+			}
+
+			uint32_t hash = _hash(p_key);
+			_insert_with_hash(hash, elem);
+			return elem;
+		}
+	}
+
+public:
+	enum {
+		MIN_CAPACITY = 16
+	};
```

Comment:
Duplicate definition, it's already set on line 70.

Reply:
Thanks! I guess until its used its difficult to get the errors to show up..